doc/man/gnunet-gns-proxy.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd October 25, 2012
.Dt GNUNET-GNS-PROXY 1
.Os
.Sh NAME
.Nm gnunet-gns-proxy
.Nd run a client-side GNS SOCKS proxy
.Sh SYNOPSIS
.Nm
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl a Ar AUTHORITY | Fl -authority= Ns Ar AUTHORITY
.Op Fl p Ar PORT | Fl -port= Ns Ar PORT
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl v | -version
.Sh DESCRIPTION
Most users will want to run this SOCKS proxy.
It can be used in combination with browsers that support the SOCKS 4a protocol.
.Pp
The proxy will perform SSL authentication of GNS names and rewrite GNS enabled HTML content.
To assert the validity of GNS names a local root CA certificate has to be generated that is used by the proxy.
Thus "gnunet-gns-proxy-setup-ca" should be executed before the first launch of this proxy or the --authority switch is used to specify an appropriate CA certificate that is already trusted by the browser.
.Bl -tag -width indent
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl a Ar AUTHORITY | Fl -authority= Ns Ar AUTHORITY
Path to a PEM CA file that contains the certificate and private key of the CA to use to assert the validity of GNS names.
The default port is specified in the configuration file for the gns service under "[gns-proxy]":
"[gns-proxy]"
PROXY_CACERT
.It Fl p Ar PORT | Fl -port= Ns Ar PORT
The port this proxy should listen on. Default is 7777.
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl v | -version
Print GNUnet version number.
.El
.\".Sh EXAMPLES
.Sh SEE ALSO
.Xr gnunet-gns 1
.sp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression
  in 0.11.0 where documentation was an on/off switch.
  New introduced: --enable-include-manpages.
  Fixed: texi2mdoc integration (--enable-texi2mdoc-generation).
  buildsystem: decrease python dependency to 3.4 from
  previously 3.7.
  -ng0

Tue Mar 12 10:06:17 CET 2019
  REST: Move back from rest-plugin into respective subsystem directories.

Mon Mar 11 02:47:55 CET 2019
  Make configure complain if no working libidn is detected. -CG

Sat Mar 9 15:58:45 2019 +0100
  REST: Config option for address bind. Defaults to localhost.
  GNS: dns2gns/gns-proxy config option for address bind. Defaults to localhost.

Sat Mar 9 01:58:22 CET 2019
  gnunet-publish now by default does not expose the creation time,
  option -d no longer exists, to enable the old behavior use -E. -CG

Sat Mar 2 14:23:10 2019 +0100

src/gns/proxy/gnunet-gns-proxy.py
#!/usr/bin/python

"""
Copyright (c) 2001 SUZUKI Hisao 

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions: 

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software. 

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""

__doc__ = """Tiny HTTP Proxy.

This module implements GET, HEAD, POST, PUT and DELETE methods
on BaseHTTPServer, and behaves as an HTTP proxy.  The CONNECT
method is also implemented experimentally, but has not been
tested yet.

Any help will be greatly appreciated.   SUZUKI Hisao
"""

__version__ = "0.2.1"

import BaseHTTPServer, select, socket, SocketServer, urlparse, re, string, os, sys

class ProxyHandler (BaseHTTPServer.BaseHTTPRequestHandler):
    __base = BaseHTTPServer.BaseHTTPRequestHandler
    __base_handle = __base.handle

    server_version = "TinyHTTPProxy/" + __version__
    rbufsize = 0                        # self.rfile Be unbuffered
    host_port = ()

    def handle(self):
        (ip, port) =  self.client_address
        if hasattr(self, 'allowed_clients') and ip not in self.allowed_clients:
            self.raw_requestline = self.rfile.readline()
            if self.parse_request(): self.send_error(403)
        else:
            self.__base_handle()

    def _connect_to(self, netloc, soc):
        i = netloc.find(':')
        to_replace = ""
        if i >= 0:
          self.host_port = netloc[:i], int(netloc[i+1:])
          if (re.match("(\w+\.)*gnunet$", self.host_port[0])):
            print 'calling gnunet-gns -a '+netloc[:i]
            auth = os.popen("gnunet-gns -a "+netloc[:i])
            lines = auth.readlines()
            if (len(lines) > 0):
              print 'result: '+lines[0].split(" ")[-1].rstrip()
              to_replace = lines[0].split(" ")[-1].rstrip()
            else:
              to_replace = "+"
        else:
          self.host_port = netloc, 80
          if (re.match("(\w+\.)*gnunet$", self.host_port[0])):
            print 'calling gnunet-gns -a '+netloc
            auth = os.popen("gnunet-gns -a "+netloc)
            lines = auth.readlines()
            if (len(lines) > 0):
              print 'result: '+lines[0].split(" ")[-1].rstrip()
              to_replace = lines[0].split(" ")[-1].rstrip()
            else:
              to_replace = "+"

        print "\t" "connect to %s:%d" % self.host_port
        try: soc.connect(self.host_port)
        except socket.error, arg:
            try: msg = arg[1]
            except: msg = arg
            self.send_error(404, msg)
            return (0, 0)
        return (1, to_replace)

    def do_CONNECT(self):
        soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            res, to_repl = self._connect_to(self.path, soc)
            if res:
                self.log_request(200)
                self.wfile.write(self.protocol_version +
                                 " 200 Connection established\r\n")
                self.wfile.write("Proxy-agent: %s\r\n" % self.version_string())
                self.wfile.write("\r\n")
                self._read_write(soc, to_repl, 300)
        finally:
            print "\t" "bye"
            soc.close()
            self.connection.close()
    
    def test_re2(self, mo):
      short = os.popen("gnunet-gns -s"+string.replace(mo.group(1), 'a href="http://', ""))
      lines = short.readlines()
      if (len(lines) < 1):
        return mo.group(1)
      elif (len(lines[0].split(" ")) > 0):
        return 'a href="http://'+lines[0].split(" ")[-1].rstrip()
      else:
        return mo.group(1)

    def shorten_zkey(self):
      return lambda mo: self.test_re2(mo)
      #return lambda mo: 'a href="http://'+os.popen("gnunet-gns -s"+string.replace(mo.group(1), 'a href="http://', "")).readlines()[0].split(" ")[-1].rstrip()
    
    def test_re(self, to_repl, mo):
      short = os.popen("gnunet-gns -s "+string.replace(mo.group(1)+to_repl, 'a href="http://', ""))
      lines = short.readlines()
      if (len(lines) < 1):
        return to_repl
      elif (len(lines[0].split(" ")) > 0):
        return 'a href="http://'+lines[0].split(" ")[-1].rstrip()
      else:
        return to_repl

    def replace_and_shorten(self, to_repl):
      return lambda mo: self.test_re(to_repl, mo)
    #  return lambda mo: 'a href="http://'+os.popen("gnunet-gns -s "+string.replace(mo.group(1)+to_repl, 'a href="http://', "")).readlines()[0].split(" ")[-1].rstrip()
    #full = string.replace(mo.group(1)+to_repl, 'a href="http://', "")
        #print 'calling gnunet-gns -s '+full
        #s = os.popen("gnunet-gns -s "+full)
        #lines = s.readlines()
        #print 'short: '+lines[0].split(" ")[-1].rstrip()
        #return 'a href="'+lines[0].split(" ")[-1].rstrip()

    def do_GET(self):
        (scm, netloc, path, params, query, fragment) = urlparse.urlparse(
            self.path, 'http')
        if scm != 'http' or fragment or not netloc:
            self.send_error(400, "bad url %s" % self.path)
            return
        soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            res, to_repl = self._connect_to(netloc, soc)
            if res:
                self.log_request()
                soc.send("%s %s %s\r\n" % (
                    self.command,
                    urlparse.urlunparse(('', '', path, params, query, '')),
                    self.request_version))
                if (re.match("(\w+\.)*gnunet$", self.headers['Host'])):
                  leho = os.popen("gnunet-gns -t LEHO -u "+self.headers['Host']).readlines()
                  if (len(leho) < 2):
                    print "Legacy hostname lookup failed!"
                  elif (len(leho) == 1):
                    print "Legacy hostname not present!"
                  else:
                    newhost = leho[1].split(" ")[-1].rstrip()
                    print "Changing Host: "+self.headers['Host']+" to "+newhost
                    self.headers['Host'] = newhost
                self.headers['Connection'] = 'close'
                del self.headers['Proxy-Connection']
                del self.headers['Accept-Encoding']
                for key_val in self.headers.items():
                    soc.send("%s: %s\r\n" % key_val)
                soc.send("\r\n")
                self._read_write(soc, to_repl)
        finally:
            print "\t" "bye"
            soc.close()
            self.connection.close()

    def _read_write(self, soc, to_repl="", max_idling=20):
        iw = [self.connection, soc]
        ow = []
        count = 0
        msg = ''
        while 1:
            count += 1
            (ins, _, exs) = select.select(iw, ow, iw, 3)
            if exs:
              break
            if ins:
                for i in ins:
                    if i is soc:
                        out = self.connection
                    else:
                        out = soc
                    data = i.recv(8192)
                    if data:
                        #try:
                          data = re.sub(r'\nAccept-Ranges: \w+', r'', data)
                          data = re.sub('(a href="http://(\w+\.)*zkey)',
                              self.shorten_zkey(), data)
                          if (re.match("(\w+\.)*gnunet$", self.host_port[0])):
                              arr = self.host_port[0].split('.')
                              arr.pop(0)
                              data = re.sub('(a href="http://(\w+\.)*)(\+)',
                                  self.replace_and_shorten(to_repl), data)
                          out.send(data)
                          count = 0
                        #except:
                        #  print "GNS exception:", sys.exc_info()[0]

            else:
                print "\t" "idle", count
                print msg
            if count == max_idling: break

    do_HEAD = do_GET
    do_POST = do_GET
    do_PUT  = do_GET
    do_DELETE=do_GET

class ThreadingHTTPServer (SocketServer.ThreadingMixIn,
                           BaseHTTPServer.HTTPServer): pass

if __name__ == '__main__':
    from sys import argv
    if argv[1:] and argv[1] in ('-h', '--help'):
        print argv[0], "[port [allowed_client_name ...]]"
    else:
        if argv[2:]:
            allowed = []
            for name in argv[2:]:
                client = socket.gethostbyname(name)
                allowed.append(client)
                print "Accept: %s (%s)" % (client, name)
            ProxyHandler.allowed_clients = allowed
            del argv[2:]
        else:
            print "Any clients will be served..."
        BaseHTTPServer.test(ProxyHandler, ThreadingHTTPServer)

